### DISCUSSION TITLE
Repost of my call for help from the Sage Forum - serious database issues getting worse

### DISCUSSION THREAD

**Discussion by LoisGNS** [OpenSageTV/Community#12] [General]
Reposting my cry for help here, since I have received no replies on the old forum. Either nobody has any solutions or someone who does just hasn't seen this yet. Getting pretty desperate! I hope it's ok to repost it here.

**Major database problems - please help!** 
________________________________________
I'm getting desperate for a solution and hope someone can help. I'll enter my current system specs at the end of this post, after I describe the problem and what I've been attempting to address it. Apologies for the length of the post, but if anyone who has a clue how to fix this could take a look, I'd be most grateful, because the issue is causing my a lot of time, trouble, frustration, etc.

Almost every day, or every few days, for the past 10 months or so and getting worse, from one to dozens of recordings get their metadata damaged in one of the following two ways:

1. The show's metadata gets changed to some random show occurring in the future
2. The show's metadata gets changed to "no data" and assigned random airing and original dates

My collection of recordings is multi-TB, and I've been using Sage since 2009, so my database is pretty old and likely very messy as a result.

_My typical processes, provided in case they could be a cause of the issue:_

I routinely convert my recordings to save space and remove any padding (using the internal conversion commands), so the resulting files don't contain the embedded metadata that I understand is included in .mpg files since (I think) Version 7. I updated from a V6 to V9 installation in 2017, and so far most of the affected shows are from before that time. A few, however, are more recent (from 2022). 

When I convert files, until the past few days (to see if it makes a difference), I have used Windows Explorer, not Sage, to delete the original files after verifying that the converted files are ok. I just started doing the deletions within Sage to see if it helps (though whether it helps will most likely not be apparent for some time, given the nature of the behavior).

Typically, each event attacks files created around the same dates - currently it's zapping files from November 2015 and May 2011. Generally, the episode ID that appears in the filename is in the same numerical range, e.g. 4826613, 4825691 are recent examples (a full filename example is "CraftinAmerica-Music-4826613-0.avi." I take this to mean that something's happening when Sage checks & updates the program guide: it finds an upcoming show with ID e.g. 4826613, and assigns whatever existing show that has that ID (if any) the metadata belonging to the upcoming show.

Edit: I also found several versions of wiz.bin.corrupt (8 of them, to be exact), dating from 6/29/22 (which is, I think, when my current round of trouble with this started - I'd had it before updating to Sage 9, but not after that until around 6/29/22. The most recent is 5/11/23. They are all of similar sizes - 75Mb to 78mb. There are also two files named Wiz.bin.PRECOMPRESSED_IDS and Wiz.bin.PRECOMPRESSED_IDS1, dated 6/29/22 and 5/10/23, respectively.

In the last day or two, I found this post:

[]([https://forums.sagetv.com/forums/sho...6&postcount=24](https://forums.sagetv.com/forums/showthread.php?p=578126&postcount=24))

and tried changing the following in .properties to see if they would help:

wizard/retain_airings_from_completed_recordings=false
wizard/last_maintenance=0

I don't understand exactly what they do, and the change probably didn't help as Sage just generated at least one new mis-meta-data recording (of type 1 above - marked as a future show). The same post I referenced had other .properties changes suggested, but I understood them even less and didn't make them for fear of causing more problems.

I've repaired the shows as they get damaged, but it's pretty time-consuming and painful. I gather that it would be less painful if I didn't run the conversions because the metadata (at least for shows recorded post 2017) would be embedded and likely picked up again in rescanning. Not converting is not very practical for my setup.

_Here's a description of my current system:_

Sage 9.2.2.904 64 bit
Java 1.8.0_191
OS: Windows 8.1.6.3
JVM Heap 815mb used/1241mb total/1241mb max
Wiz.bin is approx 75mb

_Plugins (not listing dependencies, just full plugins)_

Batch Metadata Tools - Web Interface 4.101.0
Change Channel to Record 1.1.2
Complementary Metadata Tools
Encoder Names 0.8.2
Google Weather 2.0.3 (though Weather doesn't work)
Jave Heap Monitor 1.21
Jetty Web Server 2.3. 0.17
Media Streaming Service 1.3.7.90
The Movie DB Library 4.1.0
Named Time Recording 1.36
Nielm's Sage Utilities 1.4.4
Nielm's Sage XML Info 1.4.1
OneTimeFavorite 1.02
Phoenix API 3.3.0
Phoenix Core Services 3.3.0
PlayOn for Sage TV 1.80
PushBullet notifications for Sage TV 1.2
Remove Padding on Back- to back manual recording conflicts 1.1.1
Resolve livetv conflict 1.0.0
restart sage from setup menu 1.0.3
sage for echo functions 0.05
sagetv 3 ui 9.0
SageTV Groovy Tools 1.0.4
SageTV Web Interface 3.3.1
sagex-api-Sage TV API Extensions 9.1.7.0
Sagex-services 9.1.7.0
Tuner Preroll 1.0.3
The TV DB Library 1.8.0
UPnP Browser for PlayOn 1.80
Weather Using NWS API 0.13

**What I'm thinking of doing next - any idea if it will help?**

•	Doing some thorough backups of my current setup and updating to version 9.2.6, which appears to be the most recent. If I do this, what other precautions should I take beforehand, and what will I need to do afterward? For example, will my .properties and wiz.bin get overwritten so I'll need to recopy from the backups? Will anything else get messed with and have to be re-tweaked?
•	Banging my head against the wall until I don't care any more.  
•	As a very last resort, I might try nuking my wiz.bin (after first backing it up multiply) but I suspect that would cause far more problems than I have currently given the size of my collected recordings.

**Comment by RJ-Make** [comment]
I didn't read the entire post, however I personally would start with whatever is providing/writing the metadata to you media.

**Comment by LoisGNS** [comment]
Alll of my metadata is supplied from Sage, nothing external.

**Comment by Ugadata** [comment]
Do you have the Sage Webserver installed and BMT/CMT?
This may also be doable through the properties file but ????????.
In BMT - Check if the any of the changed recordings have the top 2 options checked.
Preserve metadata & Sage recording

![image](https://github.com/OpenSageTV/Community/assets/133981660/b16214a4-94e9-4081-a143-49c4c7064ea5)


entire groups of recordings an be imported as Sage Recordings,  preserve metadata if not fixed by importing as recording would become a bit tedious.

**Reply by LoisGNS** [reply]
I do have BMT and CMT. I just finished fixing a bad batch, so have nothing to look at today - if installing the latest version, which I did the other night, didn't fix the problem, I will have some new bad recordings to check before long.  Do I want the boxes checked or unchecked?

So, for those recordings, which I'd have to search for one at a time, unless they're "no data," which I can search for as a group, what would be my next step. Typically the "no data" files have literally nothing in metadata, so I'd need to fill in manually from something saved (pretty painful, but something I've done using the web interface), or is there a way to  automate via saved .properties files (which I've been creating and archiving)? 

They are already present as recordings, just not the recordings they're supposed to be, so if the above were the example, the recording with the metadata for "Svengoolie" would actually be something else entirely. Or, if I pulled up a "no data" recording that is actually a recording of "Svengoolie," the metadata fields above would show nothing, but be listed as sage recordings. 

And is there anything I can do here to prevent the problem  from recurring over and over?

**Comment by Ugadata** [comment]
Firstly, forgive me for not paying loser attention to your original post.  It clearly indicates the Webserver and BMT.

My suggestion is more for existing recordings with correct metadata, to hopefully stop anymore recordings from having the metadata disappear.

If the recording isn't a "Sage Recording?" then (IIRC) the option to "preserve original metadata" will be present (or usable).

Admittedly, I don't really keep much in the way of old recordings.  I am typically importing movies and TV series into Sage.

Things to try (and maybe you already have?)

there are some settings in the .properties file (I can't recall what they are at this moment) for doing a recovery scan

 Move the bad files to another location that is not in any of SageTV's  "Recording Directories" or "Media Center Import Directories".  Basically, anywhere Sage can't find them.  

 Force Sage to scan for new media or wait a while until Sage does a scheduled scan.

 Move one or more files back to where Sage can see/find them, do another media scan.  My experience has been, Sage may have issues finding everything if too many files are done at once.  YMMV.

There are some posts in the old forum for the above steps and .properties file settings.

**Reply by LoisGNS** [reply]
Everything that becomes messed up is an existing Sage recording.

**Any idea where in the old forum to look for, or what to search on to find the .properties settings you mention?** 

I've been searching, but haven't found anything yet. Keep in mind, though, that my media files **do not** contain metadata, as they've been converted to save space (I get about 4:1 compression, which given the number of recordings is  important). If it's just looking for embedded metadata, that won't do any good. I wish that the conversion process (I use the one built into Sage) could add the metadata to the output file, but it doesn't. The metadata is preserved in the database, but not in the media file.

**I have had some success - but it's still quite tedious because of the number of recordings, with the following process:**

1. Check the "usual suspect" directories for any files with "created" or "modified" dates occurring in the future, and copy them to a staging area, where I also stash my saved metadata xxxx.properties files fore each one.
2. In Web server, note the problem files by checking "recordings" and noting for each what it should be (file name) vs. what Sage thinks it is (title according to the newly-erroneous metadata).
3. Also in Web server, search "no data" and  similarly note any thing I find, then copy those files from wherever they are to the staging area                                4.  Run "scan imported media." (I have this set up to run manually, as letting it run when it wants sometimes has slowed down my system)
5.  Go back into Web server, and make additional corrections to the affected media - primarily making the "airing date" match  "original air" date, so I no longer see shows that appear to be in the future. Mark in my notes file the correct date, and mark the recording as having been repaired. 
6. Go into the Sage interface and locate/play each file briefly to make sure that it is now what it should be (a necessary step because sometimes my fixes fail to take, I made a mistake, or whatever).
7. For each correctly-fixed show, delete the extra copy (with still-bad metatdata) from its original location. Mark in my notes doc that the fix has been completed.

In cases where for whatever reason that fails, use the Webserver to paste in correct metadata from my saved xxx.properties file for that show.

Either way, you can see that this is awfully tedious, and something I'd prefer to avoid having to do - especially with the high volume of files getting messed up that I've had lately.

Thanks for any further help.

**Comment by Ugadata** [comment]
Also, Since you are updating Sage, make sure you are using TheTVDB database for retrieving metadata.  IMDB doesn't work very well with Sage anymore. (Perhaps you were already aware but the metadata retrieval was changed not that long ago - again the old forum does have some info about this)

**Reply by LoisGNS** [reply]
I may have found the right place to check this setup. I found that under Phoenix Core Services, there is a setting for "TV Metadata Provider(s)," and in that field I have "tvdb." For "Movie metadata providers" it shows "tmdb, imdb." Is that  how it should be set?

**Reply by jusjoken** [reply]
Yes.  You can remove the ,imdb bit Phoenix will ignore it as that provider was removed from Phoenix many years ago.

**Reply by LoisGNS** [reply]
Thanks. So far, after updating to the latest server software version, I haven't had any new  bad-recording batches. But there are typically a few days between batches, so I'm not out of the woods. Since I already had the metadata settings as recommended, I don't know if there's anything else I can do to prevent future problems. 

The airing id seems to be part of the issue, and I found some posts in the old forum that gave a hint:

"Be sure to update the 'wizard/next_uid' setting in sage.properties when you move wiz.bin into a different setup since it holds the next valid ID."

"When a new database ID is needed, that number is used and then updated. Those ID numbers are how all of the data is related in the DB so having duplicates would mean wrong data/actors/etc associated with shows/airings. "

In my Sage.properties, I found the following line:

wizard/next_uid=787844

Oddly, though, today’s recordings have had IDs like  366253, 365334, 366499, and recent recordings that I've had to fix have had IDs like 676730, 669428, 733027, 733009, 717816, 4805790 - and various others. It's more common for the ids to all be in a similar numerical range, but the last batch, which was mostly "no data" being assigned to old recordings, had a wider range of numerical ids (by which I mean things like the number in the file name  BetterCallSaul-SeriesRecap-**673513**-0.mp4.) Would it help if I edited next_uid to some very large number? I also found "wizard/max_search_results=2500" - since I have several times 2500 recordings, should I increase this number to something that is much greater than the number of recordings I have? Or are these settings completely irrelevant, or dangerous to change?

------------------------------
User of Sage since 2009. Was **VCRUser** on the  old forum (I still have some VCRs but am using them only to slowly digitize old tapes.

**Reply by wnjj** [reply]
Wiz id’s are not related to the episode id’s you can see from the guide data.  The wiz id’s are internal serial numbers.

**Comment by techiegirl2day** [comment]
I have had issues like that off & on.  Look to corruption of the OS.  Can be directly related to HDD health.  Check your HDD. 
 The main fix is to replace the HDD & reinstall OS.  

     First try refreshing SageTV with a new profile.  IF it runs better then you know that the data is corrupted in that old profile. 
You can play whackamole or a cleanslate, your choice. Yes I know it isn't great to lose all your watch history & media lists etc. 
IF it is old it will reach a point of being too large to manage anyways, making it open to corruption.

**Reply by LoisGNS** [reply]
What do you mean by refreshing SageTV with a new profile?  A new installation in another folder? While my Sage use is old, the current computer and hard drive it's on are not so old, and are ok so far as I know (though I will look into the hdd health).

**Comment by wnjj** [comment]
If you can somehow share your wiz.bin with me, I could look into it (the ID's, etc) using my Wiz2SQL tool that turns it all into ASCII text.  Or if you wanted to give that a try yourself you can.  It's located here:  https://github.com/wnjj/wiz2sql

**Reply by LoisGNS** [reply]
I have your tool, but no good way to read the SQL. I know I tried it a few months ago and set up sqlitebrowser to try to read it, but wasn't very successful. I use MSAccess for my database operations. Do you know of a way to read your output there? I tried pasting in the SQL output by your program into an Access sql window, but it doesn't quite like the syntax. I started with the schema and after editing the first table-creation statement down to:
CREATE TABLE year (
  yearID integer, mediamask integer, tr text 
);
I was able to create that table with the correct fields.   But then I ran into more syntax issues with the data table. While with lots of hand-editing, I could conceivably get it into Access, it would be quite a slog. Do you know of a better way?

I've uploaded my wiz.bin to my Google drive & will send you a link via pm. Currently there are no known recordings with bad metadata, since I fixed the most recent batch (including 2 shows that went bad after I updated to 9.2.6, so that update doesn't seem to have fixed anything).

Thanks for any help you can offer after you've looked at my wiz.bin - or if you think I should wait until the next crop of errors apppears?

**Reply by LoisGNS** [reply]
Not sure how to send a private message here. I have a link to my wiz.bin but probably shouldn't share it with the world.

**Reply by wnjj** [reply]
Take a look at the data output in a text editor.  You can see the ID's for the objects look in the right places.  It's been awhile since I've done this.  I think you have to look at the schema output to know which column is the ID.

**Reply by jusjoken** [reply]
> Not sure how to send a private message here. I have a link to my wiz.bin but probably shouldn't share it with the world.

As far as I can tell GitHub does not have a PM replacement.  Not sure how we replace that function.

**Reply by jusjoken** [reply]
Note.. Some users, like myself, have their email address available on their profile.. So click on their user name which takes you to their profile and if there is an email shown you can email them as an alternative to PM.

**Reply by LoisGNS** [reply]
Problem is, from the data output (insert  ... into ...., etc.) I can't readily tell what looks right or wrong. 

It's clear that there are fields that need to be linked from table to table in order to then create a query that will put the pieces together in a way that makes sense. Presumably would happen more or less automagically if I had SQLite. I did find its download page, but iirc, once set up, it wasn't obvious what to do with it. Worth another look, though, since this wiz. bin problem seems to be an ongoing thing. What if I was able to find something amiss - would there be a way to fix?

My preference would be to  work in Access if I can find a way to import the data into it.  I may need to roll my own parser to get from the output SQL something that Access will process as input. There are about 2M lines of SQL code in the data, so if I could sort out the syntax differences I would first write some find/replace routines to convert those. Then I'd probably need to iterate the file & feed the sql into Access line-by-line (read, execute, repeat til done) , since it chokes when I try to feed it too much at once. 

Is there a way I can send you a DM with the link to my wiz.bin in Google Drive?

**Reply by wnjj** [reply]
So I had to refresh myself on this so looked at some old files I had laying around.  The problem with loading it into a database is that I never defined the relationships between tables so that part is not automated.  In most cases the names of the fields imply where they link but not always.  All the tool does it define the tables and pour the data into them.  So that leaves more manual methods.

If you look at the schema output you'll see the definition of all of the types of tables.  Those table definitions define the positional index for the "INSERT INTO <tablename>" statements.  For example, in an old wiz.bin I had:

INSERT INTO airing VALUES (**9707580**,1,**9707456**,32190,1541893301952,30528,0,1,5,0);

CREATE TABLE airing (
  **airingID integer DEFAULT 0,**
  mediamask integer DEFAULT 0,
  **showID integer DEFAULT 0,**
  stationID integer DEFAULT 0,
  airTime integer DEFAULT 0,
  duration integer DEFAULT 0,
  partsB integer DEFAULT 0,
  miscB integer DEFAULT 0,
  prB integer DEFAULT 0,
  persist integer DEFAULT 0
);

From that you can tell the airingID corresponds to position 1 (in the VALUES list) and is 9707580.  This airingID represents a specific airing time/date of a particular show.  The show is linked to an airing because there can be many airings for the same show.  You may then notice there are other fields with 'ID' in them.  Those map to another table.  In this example airing, the showID corresponds to position 3 and is 9707456.  By searching for "INSERT INTO show VALUES (9707456", I found this:

INSERT INTO show VALUES (9707456,1,0,5744652,"Bad Mix","A reported burglary becomes a murder investigation.",2098771,61192,0,0,0,0,0,"EP4516810015",9707456,0,1009317317120,1,9,1,0,0,0);

This is the entry that populates a record into the show table.  Notice the show is where you find the episode serial numbers that come from Schedules Direct, in this case EP451680015

CREATE TABLE show (
  showID integer DEFAULT 0,
  mediamask integer DEFAULT 0,
  duration integer DEFAULT 0,
  titleID integer DEFAULT 0,
  episodeNameStr text DEFAULT "",
  descrStr text DEFAULT "",
  categories0 integer DEFAULT 0,
  categories1 integer DEFAULT 0,
  rated integer DEFAULT 0,
  year integer DEFAULT 0,
  pr integer DEFAULT 0,
  lastwatched integer DEFAULT 0,
  forcedFirstRun integer DEFAULT 0,
  extID text DEFAULT "",
  updateCount integer DEFAULT 0,
  language integer DEFAULT 0,
  originalAirDate integer DEFAULT 0,
  seasonNum integer DEFAULT 0,
  episodeNum integer DEFAULT 0,
  cachedUnique integer DEFAULT 0,
  altEpisodeNum integer DEFAULT 0,
  showcardID integer DEFAULT 0,
  seriesID integer DEFAULT 0
);

Looking at the schema again for a show, you can see that the 4th number is the titleID, which corresponds to:

INSERT INTO title VALUES (5744652,1,"I, Detective");


Hopefully this allows you a little more insight into tracing these records.  There's much more to it, but if you are having basic corruption issues I suspect looking at airing and show will be enough to reveal something.  It'll be easiest to look specifically at the next broken airing that shows up and see what show it's pointing to and more importantly if it's pointing to a showID that another different/old airing is already pointing to.  The only way it would seem like that can happen is if one of those database ID's are shared.  They are supposed to increment forever as time goes by so they never collide.  Note that all of the ID's in wiz.bin are used across the different tables.  So a particular ID may be an airingID and the next sequential number may be an actor's ID.  This is what the wizard/next_uid entry refers to in the properties file.

Also by looking through the airing and show "INSERT INTO" entries, you should be able to locate the largest wiz ID in use.  Then search for several larger ones past those in case there are higher ones in use for other reasons.  Then make sure the next_uid is at least that high.

I'm not sure how comfortable I am sharing email publicly as I have a low tolerance for spam.  If there is an email you could provide I could contact you that way.  If you could post that or a link to your wiz.bin, I can even delete/edit that post after I grab it.  Worst case, you could email jusjoken with the link and he could forward it to me since he has my email though I wouldn't want to make a habit of that.

I'd sure like to figure out a real method to allow this sort of information back channeling but it seems GitHub removed what they used to have in favor of "openness".  Without a 3rd party system that could recognize GitHub usernames and forward email between them I don't see an easy way to avoid at least some public info.

**Reply by LoisGNS** [reply]
Thanks for the explanation of some of the tables & fields. It looks like the "airings" table includes all airings that Sage currently knows about, since several of them are shows I know I haven't recorded and don't plan to record. 

Again, if I can figure out a way to get Access to properly interpret the SQL commands to create its version of the db, I can probably sort out which fields need to be linked. I could then re-use whatever I put together to  read/convert further output from your program into Access. 

I did email Jusjoken with the link. @jusjoken, could you forward the link and my email address from which I sent it to wnjj? Then, @wnjj, if you're ok with moving this discussion to email, we can correspond that way. And, for the good of the group, if we find any solutions that might help anyone else here, I can post the results.

**Reply by wnjj** [reply]
I have your wiz.bin now and can look at it.  I discovered I had a bug since day 1 with my wiz2sql tool.  The ID's were fine but the date/time and duration fields were messed up.  I fixed that and put a new copy out here:  https://github.com/wnjj/wiz2sql/releases/latest
With that fix, you can now take the airing time and paste it into this converter: https://timestamp.online.  **You will need to remove the last 3 digits since the converter wants seconds and SageTV uses milliseconds.**

Can you tell me what your wizard/next_uid is currently in your sage.properties?  I see that you said it was 787844 above.  Is it still set to that?  I see ID's as high as 11112161 in your wiz.bin right now.  I also have to retract what I said about ID's in the filenames not being related.  They actually are wiz ID's.  That one you mentioned (BetterCallSaul-SeriesRecap-673513-0.mp4) shows up in wiz.bin as this now with its new number 730534.  The wiz ID is one more than that:

```
INSERT INTO mediaFile VALUES (730535,5,34474,"F:\FConversions\FixMeta","BetterCallSaul-SeriesRecap-730534-0.mp4",1,730534,"Mediabox4","",0,0,61,91,1,"",0,0,"f=Quicktime;dur=73950;br=1051000;[bf=vid;f=MPEG4-Video;br=953000;index=0;fps=29.97;ar=1.3333334;arn=4;ard=3;w=512;h=384;cs=yuv420p;][bf=aud;f=AAC;br=94000;index=1;sr=44100;ch=2;lang=und;]");

INSERT INTO mediaFile_file VALUES (730535,"F:\FConversions\FixMeta\BetterCallSaul-SeriesRecap-730534-0.mp4",1657414800000,1657414873950);
```

I think I will need to see a wiz.bin the next time it fails, along with the specific show with date/time of airing and what it got tangled with to see the issue.

**Reply by wnjj** [reply]
So I was able to do something really cool with Excel using the AIRING data.  Here's a plot Wiz ID of the airings by date:

![image](https://github.com/OpenSageTV/Community/assets/32583220/475b14ff-d821-47d0-bb73-4315b30a0966)

You can clearly see the two distinct ranges of wiz ID's in your data you mentioned above.  I zoomed the graph to omit the ones above 11111111 that are also in there now.

If you compare that to my setup, notice how the ID's are continuously increasing throughout time.  The horizontal bars are mass imports of photos and audio that have date/time stamps from earlier.
![image](https://github.com/OpenSageTV/Community/assets/32583220/8d437990-33ac-4931-a9b4-dcdad4a7f47d)

The point is your system looks like it's been run through the ringer.  At this point I would suggest shutting down SageTV, setting wizard/next_uid=12000000 and restarting.  Hopefully that will push all future ID's out enough.  You may still find collisions for some of the upcoming airings that are already in the schedule but hopefully after that things will clear up.

**Comment by LoisGNS** [comment]
That's consistent with the fact that the glitches seem to occur in time-bands; for example, recordings from November 2015 had numbers around 670403 that were being re-used & causing glitches, similar glitches from 2011 using numbers like 662546 or 4722250 (so two batches of numbers, I guess), numbers around 383152 started being used in 2022, and were just now starting to glitch, so that number-range was being pulled in at some point. So far the 11111111 seems ok, but for safety (next time I feel like shutting down Service) I could go higher. Do you know if there's a maximum I should stay clear of? I wasn't sure how many digits might be allowed.

**Reply by wnjj** [reply]
You should be fine with the 11,111,111 you have.  I didn't see anything other than the most recent stuff anywhere near there.  No worries about hitting the max, it's allowed to be up to 2,147,483,647.

**Comment by LoisGNS** [comment]
Hi @wnjj 
Thanks for all your help so far. Since I updated my "next_uid" to "11111111," I've had no further eruptions of bad metadata. Still a bit early to declare victory, but if it behaves for another week or two it should be good.

I managed to create Access code to import your SQL output into Access.  It's a bit clunky at the moment and has no user interface. I just run it from the code window, have input & output filenames hardcoded, etc. I plan to fix that now that the basics seem to be working. Also, I did some manual preprocessing of the schema output and should take another try at eliminating the need for that. But it gets the job done. If anyone else would find such a thing useful - especially once it's de-clunkified, I can look for a way to share it.

Now I have a full collection of populated tables,  that I can update as needed, and have created  a query that links some of them. I can't find a table that gives airing date or original airing time.  There is a field that should show it (show.originalAirDate), but only 8 records of about 60K show that field populated. Yet when I view info in  the Sage UI it knows original air dates. I haven't found actual airdates.  There are a number of tables with two fields, one called "pos," which is a sequential list of integers from 1 to n, and the other is some kind of ID field. Example: Titles_by_name: pos, TitleID. I'm not sure the various instances of the "pos" field are meant to be linked, though. Here's the first query I created, which shows me a fair bit if useful info (absent the dates mentioned):

SELECT DISTINCTROW title.str, show.originalAirDate, show.seasonNum, show.episodeNum, show.episodeNameStr, show.descrStr, airing.airingID, show.extID, mediaFile.videoDirectory, mediaFile.name
FROM mediaFile_file, ((show INNER JOIN title ON show.titleID = title.titleID) INNER JOIN airing ON show.showID = airing.showID) INNER JOIN mediaFile ON airing.airingID = mediaFile.infoAiringID
ORDER BY title.str, show.originalAirDate, show.seasonNum, show.episodeNum;

Do you have anywhere any documentation of what the various tables & fields mean? While some are self-evident, others not so much.


Thanks again!

**Comment by wnjj** [comment]
When I look in the raw data, I see many (most) of the shows with originalAirDates.  It's the 7th field from the right with the 13 digit number in it.  Is something off that is missing that field in your SQL import?

The tables with 'pos' are all just indexes.  In win.bin, the ID column was just dumped out sequentially but to make them DB-friendly I added the 'pos' to keep them in the sequential order.  These *_by_* tables are used to provided sorted lists of ID's for quick search/display. 
 They probably don't have a super-useful purpose in an SQL DB.

Unfortunately I don't have any documentation on what links to what.  I gleaned the wiz.bin format by looking in the Java code (wizard.java and all of the objects like show.java, airing.java, etc).  I generally kept the field names the same as the java variables that were handling them.  There are places where adding 'ID' may have helped with linking.  For example, the show table has "categories0/1" which are categoryID's in the category table.

**Comment by LoisGNS** [comment]
Found the problem. The info is in the raw sql output but doesn't import. The dates are 11 digits, which are longer than Access allows for long integers, which turns out to be +- 2,147,483,648. I changed the field type to string, and it came in ok; when running my imports, it just skipped over that field since it didn't fit.  I just checked, and Airing.airtime has the same issue. There are probably several other fields that should have datecodes and instead are null or 0 because of that limit.

I'll have to think about the best approach on this. Probably the easiest would be to set almost every field to string. The only one that seems to really need to be numeric is duration - which is also short enough that it behaves nicely. Then run the value trough a converter for any queries that need to use the info. 

I'll ignore the tables with "pos" in them, or maybe even filter them out during import.

**Reply by wnjj** [reply]
If you can pre-process, divide the date and time values by 1000 so they are in seconds and not milliseconds.  That is whet all of the standard converter tools want anyway.  You may also be able to coerce them into date format somehow and have them work automatically.

As an alternative, I think I'll add an option to convert those into human readable date/time as I generate the SQL.

**Reply by LoisGNS** [reply]
This is the function I made for it (using VBA), based on stuff I found online:

Function ConvertDate(UnixDate As Variant)
  UnixDate = UnixDate / 1000
  ConvertDate = Format(DateAdd("s", UnixDate, "01/01/1970"), "mm/dd/yyyy hh:mm")

End Function

Starting out with human-readable would be even better, but you'll need to re-specify the fields as date.

**Comment by LoisGNS** [comment]
So far, no more metadata mess-ups. However, despite Next_uid currently at 11209762, shows I recorded tonight have numbers like 743993 (The11thHourWithStephanieRuhle-743993-0.mpg, NBC10Newsat11p-743552-0.mpg).  Also at least one in the "correct" range (11142575), but some others all over the place, such as 388270.

Should I worry that before long I'll get more collisions? Not sure what to do about it beyond hoping for the best.

**Reply by wnjj** [reply]
Anything that was loaded from the EPG (2-3 weeks out) may have those numbers still since they got added to wiz.bin when they hit the schedule.  Give it time but I mentioned before that you may have a couple more before it flushes out.

**Reply by LoisGNS** [reply]
Thanks. I missed that angle - that the numbers were assigned by the scheduler, as opposed to when recorded. I can live with a few stragglers. It was the every-few-days dozens that were driving me to distraction.
